File: macos/ProtosMac/Sources/ProtosMacApp/AppModel.swift
import Foundation
import Observation
import SwiftUI

enum SidebarSection: String, CaseIterable, Identifiable {
    case overview = "Overview"
    case apps = "Apps"
    case provisioners = "Provisioners"
    case instances = "Instances"
    case network = "Network"
    case releases = "Releases"
    case dvc = "P2P DVC"

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .overview: "gauge.with.dots.needle.67percent"
        case .apps: "shippingbox"
        case .provisioners: "server.rack"
        case .instances: "desktopcomputer"
        case .network: "network"
        case .releases: "arrow.up.circle"
        case .dvc: "point.3.connected.trianglepath.dotted"
        }
    }
}

enum DaemonState: Equatable {
    case starting
    case running
    case failed(String)

    var title: String {
        switch self {
        case .starting:
            return "Starting"
        case .running:
            return "Running"
        case .failed:
            return "Failed"
        }
    }
}

@MainActor
@Observable
final class AppModel {
    var daemonState: DaemonState = .starting
    var selectedSection: SidebarSection? = .overview
    var message: String?
    var isBusy = false

    var userInfo: Apic_GetUserInfoResponse?
    var devices: [Apic_UserDevice] = []
    var sshKey: Apic_GetLocalSSHKeyResponse?
    var apps: [Apic_App] = []
    var supportedProvisioners: [Apic_ProvisionerType] = []
    var provisioners: [Apic_Provisioner] = []
    var instances: [Apic_CloudInstance] = []
    var networkState: Apic_NetworkState?
    var exitRoutes: [Apic_ExitRoute] = []
    var releases: [Apic_Release] = []
    var provisionerImages: [String: Apic_CloudSpecificImage] = [:] {
        didSet {
            sortedProvisionerImages = provisionerImages.sorted { $0.key < $1.key }
        }
    }
    private(set) var sortedProvisionerImages: [(key: String, value: Apic_CloudSpecificImage)] = []
    var runtimeState: Apic_RuntimeState? {
        didSet {
            runtimePeers = sortedRuntimePeers(runtimeState?.peerStatuses ?? [])
        }
    }
    private(set) var runtimePeers: [Apic_RuntimePeerStatus] = []
    var localCommits: [Apic_Commit] = []
    var outputText = ""

    @ObservationIgnored private let bridge: NativeProtosBridge
    @ObservationIgnored let api: ProtosAPI
    @ObservationIgnored private var hasStarted = false
    @ObservationIgnored private var changeWatchTask: Task<Void, Never>?
    @ObservationIgnored private var liveRefreshTask: Task<Void, Never>?
    @ObservationIgnored private var pendingLiveTables = Set<String>()
    @ObservationIgnored private var pendingRuntimeChange = false
    @ObservationIgnored private var pendingHeartbeat = false

    init(bridge: NativeProtosBridge = NativeProtosBridge()) {
        self.bridge = bridge
        self.api = ProtosAPI(bridge: bridge)
    }

    deinit {
        changeWatchTask?.cancel()
        liveRefreshTask?.cancel()
    }

    func startIfNeeded() async {
        guard !hasStarted else { return }
        hasStarted = true
        await run { [self] in
            try await self.api.start()
            self.daemonState = .running
            await self.refreshAll()
            self.startChangeWatcher()
        } onError: { [self] error in
            self.daemonState = .failed(error.localizedDescription)
        }
    }

    func refreshAll() async {
        await run { [self] in
            async let user = try? self.api.userInfo()
            async let deviceList = try? self.api.userDevices()
            async let key = try? self.api.localSSHKey()
            async let appList = try? self.api.apps()
            async let provisionerTypeList = try? self.api.supportedProvisioners()
            async let provisionerList = try? self.api.provisioners()
            async let instanceList = try? self.api.instances()
            async let network = try? self.api.networkState()
            async let routeList = try? self.api.exitRoutes()
            async let runtime = try? self.api.runtimeState()
            async let commitList = try? self.api.localCommits()

            self.userInfo = await user
            self.devices = await deviceList?.devices ?? []
            self.sshKey = await key
            self.apps = await appList?.apps ?? []
            self.supportedProvisioners = await provisionerTypeList?.provisionerTypes ?? []
            self.provisioners = await provisionerList?.provisioners ?? []
            self.instances = await instanceList?.instances ?? []
            if let networkResponse = await network, networkResponse.hasState {
                self.networkState = networkResponse.state
            } else {
                self.networkState = nil
            }
            self.exitRoutes = await routeList?.routes ?? []
            if let runtimeResponse = await runtime, runtimeResponse.hasState {
                self.runtimeState = runtimeResponse.state
            } else {
                self.runtimeState = nil
            }
            self.localCommits = await commitList?.commits ?? []
        }
    }

    func refreshNetwork(instance: String = "") async throws {
        let networkResponse = try await api.networkState(instance: instance)
        networkState = networkResponse.hasState ? networkResponse.state : nil
        let routeResponse = try await api.exitRoutes(instance: instance)
        exitRoutes = routeResponse.routes
    }

    func refreshDVC(instance: String = "") async throws {
        let runtimeResponse = try await api.runtimeState(instance: instance)
        runtimeState = runtimeResponse.hasState ? runtimeResponse.state : nil
        let commitResponse = try await api.localCommits()
        localCommits = commitResponse.commits
    }

    private func startChangeWatcher() {
        guard changeWatchTask == nil else { return }
        let stream = api.watchChanges(includeSnapshot: false, heartbeatIntervalMs: 3_000)
        changeWatchTask = Task { [weak self] in
            do {
                for try await change in stream {
                    self?.scheduleLiveRefresh(for: change)
                }
            } catch is CancellationError {
                return
            } catch {
                await MainActor.run {
                    self?.message = "Live refresh stopped: \(error.localizedDescription)"
                    self?.changeWatchTask = nil
                }
            }
        }
    }

    private func scheduleLiveRefresh(for change: Apic_WatchChangesResponse) {
        if change.reason == "heartbeat", selectedSection != .dvc {
            return
        }
        pendingLiveTables.formUnion(change.tableNames)
        pendingRuntimeChange = pendingRuntimeChange || change.runtimeChanged
        pendingHeartbeat = pendingHeartbeat || change.reason == "heartbeat"
        guard liveRefreshTask == nil else { return }
        liveRefreshTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 250_000_000)
            await self?.performPendingLiveRefresh()
        }
    }

    private func performPendingLiveRefresh() async {
        let tables = pendingLiveTables
        let runtimeChanged = pendingRuntimeChange
        let heartbeat = pendingHeartbeat
        pendingLiveTables.removeAll()
        pendingRuntimeChange = false
        pendingHeartbeat = false
        liveRefreshTask = nil

        do {
            try await refreshVisibleSection(tables: tables, runtimeChanged: runtimeChanged, heartbeat: heartbeat)
        } catch {
            message = error.localizedDescription
        }
    }

    private func refreshVisibleSection(tables: Set<String>, runtimeChanged: Bool, heartbeat: Bool) async throws {
        switch selectedSection ?? .overview {
        case .overview:
            guard !heartbeat, !runtimeChanged || !tables.isEmpty else { return }
            async let user = try? api.userInfo()
            async let deviceList = try? api.userDevices()
            async let key = try? api.localSSHKey()
            userInfo = await user
            devices = await deviceList?.devices ?? []
            sshKey = await key
        case .apps:
            guard !heartbeat, !runtimeChanged || !tables.isEmpty else { return }
            apps = try await api.apps().apps
        case .provisioners:
            guard !heartbeat, !runtimeChanged || !tables.isEmpty else { return }
            async let supported = try? api.supportedProvisioners()
            async let existing = try? api.provisioners()
            supportedProvisioners = await supported?.provisionerTypes ?? []
            provisioners = await existing?.provisioners ?? []
        case .instances:
            guard !heartbeat, !runtimeChanged || !tables.isEmpty else { return }
            instances = try await api.instances().instances
        case .network:
            guard !heartbeat else { return }
            try await refreshNetwork()
        case .releases:
            return
        case .dvc:
            if heartbeat || runtimeChanged || !tables.isEmpty {
                try await refreshDVC()
            }
        }
    }

    func run(_ operation: @escaping () async throws -> Void) async {
        await run(operation, onError: nil)
    }

    func run(
        _ operation: @escaping () async throws -> Void,
        onError: ((Error) -> Void)?
    ) async {
        isBusy = true
        message = nil
        do {
            try await operation()
        } catch {
            message = error.localizedDescription
            onError?(error)
        }
        isBusy = false
    }
}

private func sortedRuntimePeers(_ peers: [Apic_RuntimePeerStatus]) -> [Apic_RuntimePeerStatus] {
    peers.sorted {
        if $0.connected != $1.connected {
            return $0.connected && !$1.connected
        }
        return $0.peerID.localizedStandardCompare($1.peerID) == .orderedAscending
    }
}

extension String {
    var nonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    var credentialPairs: [String: String] {
        split(whereSeparator: { $0 == "\n" || $0 == "," })
            .reduce(into: [String: String]()) { result, rawPair in
                let pair = rawPair.split(separator: "=", maxSplits: 1).map {
                    String($0).trimmingCharacters(in: .whitespacesAndNewlines)
                }
                guard pair.count == 2, !pair[0].isEmpty else { return }
                result[pair[0]] = pair[1]
            }
    }

    var routeCIDRs: [String] {
        split(whereSeparator: { $0 == "\n" || $0 == "," })
            .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}

File: clients/shared/flutter/lib/src/app_model.dart
import 'dart:async';
import 'dart:io' show Platform;

import 'package:flutter/material.dart';

import 'generated/apic/proto/apic.pb.dart' as pb;
import 'native/protos_bridge.dart';
import 'native/protos_tunnel_bridge.dart';
import 'protos_api.dart';
import 'text_helpers.dart';

const _iosTunnelBuildEnabled = bool.fromEnvironment(
  'PROTOS_IOS_TUNNEL',
  defaultValue: true,
);

enum SidebarSection {
  overview('Overview', Icons.speed_outlined),
  apps('Apps', Icons.inventory_2_outlined),
  provisioners('Provisioners', Icons.dns_outlined),
  instances('Instances', Icons.computer_outlined),
  tasks('Tasks', Icons.task_alt_outlined),
  network('Network', Icons.hub_outlined),
  releases('Releases', Icons.arrow_circle_up_outlined),
  dvc('P2P DVC', Icons.account_tree_outlined),
  status('Status', Icons.monitor_heart_outlined);

  const SidebarSection(this.label, this.icon);

  final String label;
  final IconData icon;
}

enum DaemonStatus { starting, running, failed }

class DaemonState {
  const DaemonState._(this.status, [this.error]);

  static const starting = DaemonState._(DaemonStatus.starting);
  static const running = DaemonState._(DaemonStatus.running);

  factory DaemonState.failed(String error) {
    return DaemonState._(DaemonStatus.failed, error);
  }

  final DaemonStatus status;
  final String? error;

  String get title {
    return switch (status) {
      DaemonStatus.starting => 'Starting',
      DaemonStatus.running => 'Running',
      DaemonStatus.failed => 'Failed',
    };
  }

  bool get isRunning => status == DaemonStatus.running;
}

class AppModel extends ChangeNotifier {
  AppModel({NativeProtosBridge? bridge, ProtosTunnelBridge? tunnelBridge})
    : bridge = bridge ?? NativeProtosBridge(),
      tunnelBridge = tunnelBridge ?? ProtosTunnelBridge() {
    api = ProtosApi(this.bridge);
  }

  final NativeProtosBridge bridge;
  final ProtosTunnelBridge tunnelBridge;
  late final ProtosApi api;

  DaemonState daemonState = DaemonState.starting;
  SidebarSection _selectedSection = SidebarSection.overview;
  String? message;
  var isBusy = false;
  var needsInitialization = false;

  pb.GetUserInfoResponse? userInfo;
  List<pb.UserDevice> devices = [];
  pb.GetLocalSSHKeyResponse? sshKey;
  List<pb.App> apps = [];
  List<pb.ProvisionerType> supportedProvisioners = [];
  List<pb.Provisioner> provisioners = [];
  List<pb.CloudInstance> instances = [];
  List<pb.Task> tasks = [];
  String selectedTaskId = '';
  List<pb.TaskEvent> selectedTaskEvents = [];
  pb.NetworkState? networkState;
  List<pb.ExitRoute> exitRoutes = [];
  List<pb.Release> releases = [];
  Map<String, pb.CloudSpecificImage> provisionerImages = {};
  List<MapEntry<String, pb.CloudSpecificImage>> sortedProvisionerImages = [];
  pb.RuntimeState? runtimeState;
  List<pb.RuntimePeerStatus> runtimePeers = [];
  List<pb.Commit> localCommits = [];
  pb.SystemStatus? systemStatus;
  MobileTunnelStatus mobileTunnelStatus =
      const MobileTunnelStatus.unsupported();
  pb.GetInstanceDeployOptionsResponse? instanceDeployOptions;
  String outputText = '';

  var _hasStarted = false;
  var _busyCount = 0;
  StreamSubscription<pb.WatchChangesResponse>? _changeWatchSubscription;
  Timer? _liveRefreshTimer;
  Timer? _initializationProbeTimer;
  final _pendingLiveTables = <String>{};
  var _pendingRuntimeChange = false;
  var _pendingHeartbeat = false;

  SidebarSection get selectedSection => _selectedSection;
  bool get supportsNetwork => !Platform.isIOS;
  bool get supportsHostAgent => !Platform.isIOS;
  bool get supportsMobileTunnel =>
      Platform.isIOS && _iosTunnelBuildEnabled && tunnelBridge.isSupported;

  set selectedSection(SidebarSection section) {
    if (needsInitialization && section != SidebarSection.overview) {
      return;
    }
    if (_selectedSection == section) {
      return;
    }
    _selectedSection = section;
    notifyListeners();
    if (daemonState.isRunning) {
      switch (section) {
        case SidebarSection.dvc:
          unawaited(run(() => refreshDvc(notify: false)));
        case SidebarSection.status:
          unawaited(run(() => refreshStatus(notify: false)));
        case SidebarSection.tasks:
          unawaited(run(() => refreshTasks(notify: false)));
        case SidebarSection.overview:
        case SidebarSection.apps:
        case SidebarSection.provisioners:
        case SidebarSection.instances:
        case SidebarSection.network:
        case SidebarSection.releases:
          break;
      }
    }
  }

  Future<void> startIfNeeded() async {
    if (_hasStarted) {
      return;
    }
    _hasStarted = true;
    await run(
      () async {
        await api.start();
        daemonState = DaemonState.running;
        try {
          userInfo = await api.userInfo();
        } catch (_) {
          _enterInitializationMode();
          return;
        }
        needsInitialization = false;
        await refreshAll(notify: false);
        _startChangeWatcher();
      },
      onError: (error) {
        daemonState = DaemonState.failed(error.toString());
      },
    );
  }

  Future<void> initializeUser({
    required String username,
    required String name,
    required String organization,
  }) async {
    await api.initUser(
      username: username,
      name: name,
      organization: organization,
    );
    await _leaveInitializationMode();
  }

  Future<void> checkInitialization() async {
    userInfo = await api.userInfo();
    await _leaveInitializationMode();
  }

  Future<void> refreshAll({bool notify = true}) async {
    final user = _optional(api.userInfo());
    final deviceList = _optional(api.userDevices());
    final key = _optional(api.localSshKey());
    final appList = _optional(api.apps());
    final provisionerTypeList = _optional(api.supportedProvisioners());
    final provisionerList = _optional(api.provisioners());
    final instanceList = _optional(api.instances());
    final taskList = _optional(api.tasks(maxResults: 200));
    final network = supportsNetwork
        ? _optional(api.networkState())
        : Future<pb.GetNetworkStateResponse?>.value();
    final routeList = supportsNetwork
        ? _optional(api.exitRoutes())
        : Future<pb.GetExitRoutesResponse?>.value();
    final runtime = _optional(api.runtimeState());
    final commitList = _optional(api.localCommits());
    final status = _optional(api.systemStatus());
    final tunnelStatus = supportsMobileTunnel
        ? _optional(tunnelBridge.tunnelStatus())
        : Future<MobileTunnelStatus?>.value();

    userInfo = await user;
    devices = (await deviceList)?.devices.toList(growable: false) ?? [];
    sshKey = await key;
    apps = (await appList)?.apps.toList(growable: false) ?? [];
    supportedProvisioners =
        (await provisionerTypeList)?.provisionerTypes.toList(growable: false) ??
        [];
    provisioners =
        (await provisionerList)?.provisioners.toList(growable: false) ?? [];
    instances = (await instanceList)?.instances.toList(growable: false) ?? [];
    tasks = (await taskList)?.tasks.toList(growable: false) ?? [];
    _pruneSelectedTask();
    final networkResponse = await network;
    networkState = networkResponse?.hasState() == true
        ? networkResponse!.state
        : null;
    exitRoutes = (await routeList)?.routes.toList(growable: false) ?? [];
    final runtimeResponse = await runtime;
    _setRuntimeState(
      runtimeResponse?.hasState() == true ? runtimeResponse!.state : null,
    );
    localCommits = (await commitList)?.commits.toList(growable: false) ?? [];
    final statusResponse = await status;
    systemStatus = statusResponse?.hasStatus() == true
        ? statusResponse!.status
        : null;
    mobileTunnelStatus =
        await tunnelStatus ?? const MobileTunnelStatus.unsupported();
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshOverview({bool notify = true}) async {
    final user = _optional(api.userInfo());
    final deviceList = _optional(api.userDevices());
    final key = _optional(api.localSshKey());
    userInfo = await user;
    devices = (await deviceList)?.devices.toList(growable: false) ?? [];
    sshKey = await key;
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshApps({bool notify = true}) async {
    apps = (await api.apps()).apps.toList(growable: false);
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshProvisioners({bool notify = true}) async {
    final supported = _optional(api.supportedProvisioners());
    final existing = _optional(api.provisioners());
    supportedProvisioners =
        (await supported)?.provisionerTypes.toList(growable: false) ?? [];
    provisioners = (await existing)?.provisioners.toList(growable: false) ?? [];
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshNetwork({
    String instance = '',
    bool notify = true,
  }) async {
    if (!supportsNetwork) {
      networkState = null;
      exitRoutes = [];
      if (supportsMobileTunnel) {
        await refreshMobileTunnelStatus(notify: false);
      }
      if (notify) {
        notifyListeners();
      }
      return;
    }
    final networkResponse = await api.networkState(instance: instance);
    networkState = networkResponse.hasState() ? networkResponse.state : null;
    final routeResponse = await api.exitRoutes(instance: instance);
    exitRoutes = routeResponse.routes.toList(growable: false);
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshInstances({bool notify = true}) async {
    instances = (await api.instances()).instances.toList(growable: false);
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshTasks({bool notify = true}) async {
    final response = await api.tasks(maxResults: 200);
    tasks = response.tasks.toList(growable: false);
    _pruneSelectedTask();
    if (selectedTaskId.nonEmpty != null) {
      await refreshSelectedTaskDetail(notify: false);
    }
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> selectTask(String id) async {
    selectedTaskId = id.trim();
    selectedTaskEvents = [];
    notifyListeners();
    if (selectedTaskId.nonEmpty == null) {
      return;
    }
    await refreshSelectedTaskDetail();
  }

  Future<void> refreshSelectedTaskDetail({bool notify = true}) async {
    final id = selectedTaskId.nonEmpty;
    if (id == null) {
      selectedTaskEvents = [];
      if (notify) {
        notifyListeners();
      }
      return;
    }
    final response = await api.task(id, includeEvents: true);
    if (response.hasTask()) {
      _upsertTask(response.task);
    }
    selectedTaskEvents = response.events.toList(growable: false);
    if (notify) {
      notifyListeners();
    }
  }

  pb.Task? get selectedTask {
    final id = selectedTaskId.nonEmpty;
    if (id == null) {
      return null;
    }
    for (final task in tasks) {
      if (task.id == id) {
        return task;
      }
    }
    return null;
  }

  Future<void> refreshInstanceDeployOptions({
    String provisioner = '',
    String location = '',
    bool notify = true,
  }) async {
    instanceDeployOptions = await api.instanceDeployOptions(
      provisioner: provisioner,
      location: location,
    );
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshDvc({String instance = '', bool notify = true}) async {
    final runtimeResponse = await api.runtimeState(instance: instance);
    _setRuntimeState(runtimeResponse.hasState() ? runtimeResponse.state : null);
    final commitResponse = await api.localCommits();
    localCommits = commitResponse.commits.toList(growable: false);
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshStatus({bool notify = true}) async {
    final responseFuture = api.systemStatus();
    final tunnelStatus = supportsMobileTunnel
        ? _optional(tunnelBridge.tunnelStatus())
        : Future<MobileTunnelStatus?>.value();
    final response = await responseFuture;
    systemStatus = response.hasStatus() ? response.status : null;
    mobileTunnelStatus =
        await tunnelStatus ?? const MobileTunnelStatus.unsupported();
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> refreshMobileTunnelStatus({bool notify = true}) async {
    mobileTunnelStatus = supportsMobileTunnel
        ? await tunnelBridge.tunnelStatus()
        : const MobileTunnelStatus.unsupported();
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> installOrUpdateMobileTunnel({
    required String instance,
    String dnsServer = '',
    List<String> cidrs = const [],
  }) async {
    if (!supportsMobileTunnel) {
      throw UnsupportedError('The Protos mobile tunnel is iOS-only.');
    }
    final response = await api.mobileTunnelConfig(
      instance: instance,
      dnsServer: dnsServer,
      cidrs: cidrs,
    );
    if (!response.hasConfig()) {
      throw StateError('The embedded core did not return a tunnel config.');
    }
    await tunnelBridge.installOrUpdateTunnel(response.config);
    mobileTunnelStatus = await tunnelBridge.tunnelStatus();
    outputText =
        'Mobile tunnel profile installed for ${response.config.instanceName}.';
  }

  Future<void> startMobileTunnel() async {
    if (!supportsMobileTunnel) {
      throw UnsupportedError('The Protos mobile tunnel is iOS-only.');
    }
    await tunnelBridge.startTunnel();
    mobileTunnelStatus = await tunnelBridge.tunnelStatus();
  }

  Future<void> stopMobileTunnel() async {
    if (!supportsMobileTunnel) {
      throw UnsupportedError('The Protos mobile tunnel is iOS-only.');
    }
    await tunnelBridge.stopTunnel();
    mobileTunnelStatus = await tunnelBridge.tunnelStatus();
  }

  void setProvisionerImages(
    Map<String, pb.CloudSpecificImage> images, {
    bool notify = true,
  }) {
    provisionerImages = images;
    sortedProvisionerImages = images.entries.toList(growable: false)
      ..sort((a, b) => a.key.compareTo(b.key));
    if (notify) {
      notifyListeners();
    }
  }

  Future<void> run(
    Future<void> Function() operation, {
    void Function(Object error)? onError,
  }) async {
    final notifyAtStart = !isBusy || message != null;
    _busyCount += 1;
    isBusy = true;
    message = null;
    if (notifyAtStart) {
      notifyListeners();
    }
    try {
      await operation();
    } catch (error) {
      debugPrint('Protos operation failed: $error');
      message = error.toString();
      onError?.call(error);
    } finally {
      _setBusy(false);
    }
  }

  void clearMessage() {
    message = null;
    notifyListeners();
  }

  @override
  void dispose() {
    _changeWatchSubscription?.cancel();
    _liveRefreshTimer?.cancel();
    _initializationProbeTimer?.cancel();
    unawaited(api.stop().whenComplete(bridge.dispose));
    super.dispose();
  }

  void _setRuntimeState(pb.RuntimeState? state) {
    runtimeState = state;
    runtimePeers = _sortedRuntimePeers(state?.peerStatuses ?? const []);
  }

  void _setBusy(bool busy) {
    _busyCount += busy ? 1 : -1;
    if (_busyCount < 0) {
      _busyCount = 0;
    }
    final next = _busyCount > 0;
    if (isBusy == next) {
      return;
    }
    isBusy = next;
    notifyListeners();
  }

  void _startChangeWatcher() {
    if (needsInitialization) {
      return;
    }
    if (_changeWatchSubscription != null) {
      return;
    }
    _changeWatchSubscription = api
        .watchChanges(heartbeatIntervalMs: 3000)
        .listen(
          _scheduleLiveRefresh,
          onError: (Object error) {
            message = 'Live refresh stopped: $error';
            _changeWatchSubscription = null;
            notifyListeners();
          },
          onDone: () {
            _changeWatchSubscription = null;
            notifyListeners();
          },
        );
  }

  void _scheduleLiveRefresh(pb.WatchChangesResponse change) {
    if (needsInitialization) {
      return;
    }
    if (change.reason == 'heartbeat' &&
        selectedSection != SidebarSection.dvc &&
        selectedSection != SidebarSection.status &&
        selectedSection != SidebarSection.tasks) {
      return;
    }
    _pendingLiveTables.addAll(change.tableNames);
    _pendingRuntimeChange = _pendingRuntimeChange || change.runtimeChanged;
    _pendingHeartbeat = _pendingHeartbeat || change.reason == 'heartbeat';
    _liveRefreshTimer ??= Timer(
      const Duration(milliseconds: 500),
      () => unawaited(_performPendingLiveRefresh()),
    );
  }

  Future<void> _performPendingLiveRefresh() async {
    if (needsInitialization) {
      return;
    }
    final tables = Set<String>.from(_pendingLiveTables);
    final runtimeChanged = _pendingRuntimeChange;
    final heartbeat = _pendingHeartbeat;
    _pendingLiveTables.clear();
    _pendingRuntimeChange = false;
    _pendingHeartbeat = false;
    _liveRefreshTimer = null;

    try {
      await _refreshVisibleSection(
        tables: tables,
        runtimeChanged: runtimeChanged,
        heartbeat: heartbeat,
      );
    } catch (error) {
      message = error.toString();
      notifyListeners();
    }
  }

  Future<void> _refreshVisibleSection({
    required Set<String> tables,
    required bool runtimeChanged,
    required bool heartbeat,
  }) async {
    if (needsInitialization) {
      return;
    }
    var refreshed = false;
    switch (selectedSection) {
      case SidebarSection.overview:
        if (heartbeat || (runtimeChanged && tables.isEmpty)) {
          return;
        }
        await refreshOverview(notify: false);
        refreshed = true;
      case SidebarSection.apps:
        if (heartbeat || (runtimeChanged && tables.isEmpty)) {
          return;
        }
        await refreshApps(notify: false);
        refreshed = true;
      case SidebarSection.provisioners:
        if (heartbeat || (runtimeChanged && tables.isEmpty)) {
          return;
        }
        await refreshProvisioners(notify: false);
        refreshed = true;
      case SidebarSection.instances:
        if (heartbeat || (runtimeChanged && tables.isEmpty)) {
          return;
        }
        await refreshInstances(notify: false);
        refreshed = true;
      case SidebarSection.tasks:
        if (heartbeat ||
            runtimeChanged ||
            tables.contains('tasks') ||
            tables.contains('task_events')) {
          await refreshTasks(notify: false);
          refreshed = true;
        }
      case SidebarSection.network:
        if (heartbeat || !supportsNetwork) {
          return;
        }
        await refreshNetwork(notify: false);
        refreshed = true;
      case SidebarSection.releases:
        return;
      case SidebarSection.dvc:
        if (heartbeat || runtimeChanged || tables.isNotEmpty) {
          await refreshDvc(notify: false);
          refreshed = true;
        }
      case SidebarSection.status:
        if (heartbeat || runtimeChanged || tables.isNotEmpty) {
          await refreshStatus(notify: false);
          refreshed = true;
        }
    }
    if (refreshed) {
      notifyListeners();
    }
  }

  void _enterInitializationMode() {
    needsInitialization = true;
    _selectedSection = SidebarSection.overview;
    message = null;
    _clearLoadedData();
    _stopChangeWatcher();
    _startInitializationProbe();
    notifyListeners();
  }

  Future<void> _leaveInitializationMode() async {
    needsInitialization = false;
    _selectedSection = SidebarSection.overview;
    _initializationProbeTimer?.cancel();
    _initializationProbeTimer = null;
    await refreshAll(notify: false);
    _startChangeWatcher();
    notifyListeners();
  }

  void _clearLoadedData() {
    userInfo = null;
    devices = [];
    sshKey = null;
    apps = [];
    supportedProvisioners = [];
    provisioners = [];
    instances = [];
    tasks = [];
    selectedTaskId = '';
    selectedTaskEvents = [];
    networkState = null;
    exitRoutes = [];
    releases = [];
    provisionerImages = {};
    sortedProvisionerImages = [];
    runtimeState = null;
    runtimePeers = [];
    localCommits = [];
    systemStatus = null;
    mobileTunnelStatus = const MobileTunnelStatus.unsupported();
    instanceDeployOptions = null;
    outputText = '';
  }

  void _stopChangeWatcher() {
    unawaited(_changeWatchSubscription?.cancel());
    _changeWatchSubscription = null;
    _liveRefreshTimer?.cancel();
    _liveRefreshTimer = null;
    _pendingLiveTables.clear();
    _pendingRuntimeChange = false;
    _pendingHeartbeat = false;
  }

  void _startInitializationProbe() {
    _initializationProbeTimer ??= Timer.periodic(
      const Duration(seconds: 3),
      (_) => unawaited(_checkExternalInitialization()),
    );
  }

  Future<void> _checkExternalInitialization() async {
    if (!needsInitialization || isBusy || !daemonState.isRunning) {
      return;
    }
    try {
      userInfo = await api.userInfo();
      await _leaveInitializationMode();
    } catch (_) {
      // Stay in the guided setup flow until initialization completes.
    }
  }

  void _pruneSelectedTask() {
    final id = selectedTaskId.nonEmpty;
    if (id == null) {
      selectedTaskEvents = [];
      return;
    }
    if (tasks.any((task) => task.id == id)) {
      return;
    }
    selectedTaskId = '';
    selectedTaskEvents = [];
  }

  void _upsertTask(pb.Task task) {
    final index = tasks.indexWhere((existing) => existing.id == task.id);
    if (index < 0) {
      tasks = [task, ...tasks];
      return;
    }
    final next = tasks.toList(growable: true);
    next[index] = task;
    tasks = next;
  }
}

Future<T?> _optional<T>(Future<T> future) async {
  try {
    return await future;
  } catch (_) {
    return null;
  }
}

List<pb.RuntimePeerStatus> _sortedRuntimePeers(
  Iterable<pb.RuntimePeerStatus> peers,
) {
  final sorted = peers.toList(growable: false);
  sorted.sort((a, b) {
    if (a.connected != b.connected) {
      return a.connected ? -1 : 1;
    }
    return a.peerId.compareTo(b.peerId);
  });
  return sorted;
}
